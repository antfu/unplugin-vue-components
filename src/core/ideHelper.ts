import { dirname, relative, isAbsolute } from 'path'
import { promises as fs, existsSync } from 'fs'
import { notNullish, slash } from '@antfu/utils'
import { Context } from './context'
import { getVueVersion } from './options'
import { getTransformedPath } from './utils'

export async function generateIdeHelper(ctx: Context, root: string, filepath: string) {
  const imports: Record<string, string> = Object.fromEntries(
    Object.values({
      ...ctx.componentNameMap,
      ...ctx.componentCustomMap,
    })
      .map(({ path, name, importName }) => {
        if (!name)
          return undefined
        path = getTransformedPath(path, ctx)
        const related = isAbsolute(path)
          ? `./${relative(dirname(filepath), path)}`
          : path

        let entry = 'import '
        if (importName)
          entry += `{ ${importName} as ${name} }`
        else
          entry += name
        entry += ` from '${slash(related)}';`
        return [name, entry]
      })
      .filter(notNullish),
  )

  if (!Object.keys(imports).length)
    return

  const originalContent = existsSync(filepath) ? await fs.readFile(filepath, 'utf-8') : ''

  const lines = Object.entries(imports)
    .sort((a, b) => a[0].localeCompare(b[0]))

  let code = `// generated by unplugin-vue-components
// We suggest you to NOT commit this file into source control
// Read more: https://github.com/antfu/unplugin-vue-components/issues/135
`

  if (getVueVersion() === 'vue3') {
    // @see https://youtrack.jetbrains.com/issue/WEB-48239
    code += `import { createApp } from "vue";
    
${lines.map(line => line[1]).join('\n')}

const app = createApp({});

const Vue = app

${lines.map(line => `Vue.component('${line[0]}', ${line[0]});`).join('\n')}

app.mount("body");

`
  }
  else {
    code += `import Vue from "vue";
    
${lines.map(line => line[1]).join('\n')}

${lines.map(line => `Vue.component('${line[0]}', ${line[0]});\nVue.component('Lazy${line[0]}', ${line[0]});`).join('\n')}

`
  }

  if (code !== originalContent)
    await fs.writeFile(filepath, code, 'utf-8')
}
