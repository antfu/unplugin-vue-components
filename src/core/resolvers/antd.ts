// ref: https://github.com/antfu/unplugin-vue-components/blob/main/src/core/resolvers/antdv.ts

import type { ComponentResolverObject } from '../../types';

interface AntdExported {
  local: string;
  exported: string;
  source: string;
  styleDir?: string | null | undefined | void;
}

const antdExportedMap: { [x: string]: AntdExported } = {
  Affix: {
    local: 'default',
    exported: 'Affix',
    source: 'affix/index',
    styleDir: 'affix',
  },
  Anchor: {
    local: 'default',
    exported: 'Anchor',
    source: 'anchor/index',
    styleDir: 'anchor',
  },
  AutoComplete: {
    local: 'default',
    exported: 'AutoComplete',
    source: 'auto-complete/index',
    styleDir: 'auto-complete',
  },
  Alert: {
    local: 'default',
    exported: 'Alert',
    source: 'alert/index',
    styleDir: 'alert',
  },
  Avatar: {
    local: 'default',
    exported: 'Avatar',
    source: 'avatar/index',
    styleDir: 'avatar',
  },
  BackTop: {
    local: 'default',
    exported: 'BackTop',
    source: 'back-top/index',
    styleDir: 'back-top',
  },
  Badge: {
    local: 'default',
    exported: 'Badge',
    source: 'badge/index',
    styleDir: 'badge',
  },
  Base: {
    local: 'default',
    exported: 'Base',
    source: 'base/index',
    styleDir: 'base',
  },
  Breadcrumb: {
    local: 'default',
    exported: 'Breadcrumb',
    source: 'breadcrumb/index',
    styleDir: 'breadcrumb',
  },
  Button: {
    local: 'default',
    exported: 'Button',
    source: 'button/index',
    styleDir: 'button',
  },
  Calendar: {
    local: 'default',
    exported: 'Calendar',
    source: 'calendar/index',
    styleDir: 'calendar',
  },
  Card: {
    local: 'default',
    exported: 'Card',
    source: 'card/index',
    styleDir: 'card',
  },
  Collapse: {
    local: 'default',
    exported: 'Collapse',
    source: 'collapse/index',
    styleDir: 'collapse',
  },
  Carousel: {
    local: 'default',
    exported: 'Carousel',
    source: 'carousel/index',
    styleDir: 'carousel',
  },
  Cascader: {
    local: 'default',
    exported: 'Cascader',
    source: 'cascader/index',
    styleDir: 'cascader',
  },
  Checkbox: {
    local: 'default',
    exported: 'Checkbox',
    source: 'checkbox/index',
    styleDir: 'checkbox',
  },
  Col: {
    local: 'default',
    exported: 'Col',
    source: 'col/index',
    styleDir: 'col',
  },
  DatePicker: {
    local: 'default',
    exported: 'DatePicker',
    source: 'date-picker/index',
    styleDir: 'date-picker',
  },
  Divider: {
    local: 'default',
    exported: 'Divider',
    source: 'divider/index',
    styleDir: 'divider',
  },
  Dropdown: {
    local: 'default',
    exported: 'Dropdown',
    source: 'dropdown/index',
    styleDir: 'dropdown',
  },
  Form: {
    local: 'default',
    exported: 'Form',
    source: 'form/index',
    styleDir: 'form',
  },
  FormModel: {
    local: 'default',
    exported: 'FormModel',
    source: 'form-model/index',
    styleDir: 'form-model',
  },
  Icon: {
    local: 'default',
    exported: 'Icon',
    source: 'icon/index',
    styleDir: 'icon',
  },
  Input: {
    local: 'default',
    exported: 'Input',
    source: 'input/index',
    styleDir: 'input',
  },
  InputNumber: {
    local: 'default',
    exported: 'InputNumber',
    source: 'input-number/index',
    styleDir: 'input-number',
  },
  Layout: {
    local: 'default',
    exported: 'Layout',
    source: 'layout/index',
    styleDir: 'layout',
  },
  List: {
    local: 'default',
    exported: 'List',
    source: 'list/index',
    styleDir: 'list',
  },
  LocaleProvider: {
    local: 'default',
    exported: 'LocaleProvider',
    source: 'locale-provider/index',
    styleDir: 'locale-provider',
  },
  message: {
    local: 'default',
    exported: 'message',
    source: 'message/index',
    styleDir: 'message',
  },
  Menu: {
    local: 'default',
    exported: 'Menu',
    source: 'menu/index',
    styleDir: 'menu',
  },
  Mentions: {
    local: 'default',
    exported: 'Mentions',
    source: 'mentions/index',
    styleDir: 'mentions',
  },
  Modal: {
    local: 'default',
    exported: 'Modal',
    source: 'modal/index',
    styleDir: 'modal',
  },
  notification: {
    local: 'default',
    exported: 'notification',
    source: 'notification/index',
    styleDir: 'notification',
  },
  Pagination: {
    local: 'default',
    exported: 'Pagination',
    source: 'pagination/index',
    styleDir: 'pagination',
  },
  Popconfirm: {
    local: 'default',
    exported: 'Popconfirm',
    source: 'popconfirm/index',
    styleDir: 'popconfirm',
  },
  Popover: {
    local: 'default',
    exported: 'Popover',
    source: 'popover/index',
    styleDir: 'popover',
  },
  Progress: {
    local: 'default',
    exported: 'Progress',
    source: 'progress/index',
    styleDir: 'progress',
  },
  Radio: {
    local: 'default',
    exported: 'Radio',
    source: 'radio/index',
    styleDir: 'radio',
  },
  Rate: {
    local: 'default',
    exported: 'Rate',
    source: 'rate/index',
    styleDir: 'rate',
  },
  Row: {
    local: 'default',
    exported: 'Row',
    source: 'row/index',
    styleDir: 'row',
  },
  Select: {
    local: 'default',
    exported: 'Select',
    source: 'select/index',
    styleDir: 'select',
  },
  Slider: {
    local: 'default',
    exported: 'Slider',
    source: 'slider/index',
    styleDir: 'slider',
  },
  Spin: {
    local: 'default',
    exported: 'Spin',
    source: 'spin/index',
    styleDir: 'spin',
  },
  Statistic: {
    local: 'default',
    exported: 'Statistic',
    source: 'statistic/index',
    styleDir: 'statistic',
  },
  Steps: {
    local: 'default',
    exported: 'Steps',
    source: 'steps/index',
    styleDir: 'steps',
  },
  Switch: {
    local: 'default',
    exported: 'Switch',
    source: 'switch/index',
    styleDir: 'switch',
  },
  Table: {
    local: 'default',
    exported: 'Table',
    source: 'table/index',
    styleDir: 'table',
  },
  Transfer: {
    local: 'default',
    exported: 'Transfer',
    source: 'transfer/index',
    styleDir: 'transfer',
  },
  Tree: {
    local: 'default',
    exported: 'Tree',
    source: 'tree/index',
    styleDir: 'tree',
  },
  TreeSelect: {
    local: 'default',
    exported: 'TreeSelect',
    source: 'tree-select/index',
    styleDir: 'tree-select',
  },
  Tabs: {
    local: 'default',
    exported: 'Tabs',
    source: 'tabs/index',
    styleDir: 'tabs',
  },
  Tag: {
    local: 'default',
    exported: 'Tag',
    source: 'tag/index',
    styleDir: 'tag',
  },
  TimePicker: {
    local: 'default',
    exported: 'TimePicker',
    source: 'time-picker/index',
    styleDir: 'time-picker',
  },
  Timeline: {
    local: 'default',
    exported: 'Timeline',
    source: 'timeline/index',
    styleDir: 'timeline',
  },
  Tooltip: {
    local: 'default',
    exported: 'Tooltip',
    source: 'tooltip/index',
    styleDir: 'tooltip',
  },
  Upload: {
    local: 'default',
    exported: 'Upload',
    source: 'upload/index',
    styleDir: 'upload',
  },
  version: {
    local: 'default',
    exported: 'version',
    source: 'version/index',
    styleDir: 'version',
  },
  Drawer: {
    local: 'default',
    exported: 'Drawer',
    source: 'drawer/index',
    styleDir: 'drawer',
  },
  Skeleton: {
    local: 'default',
    exported: 'Skeleton',
    source: 'skeleton/index',
    styleDir: 'skeleton',
  },
  Comment: {
    local: 'default',
    exported: 'Comment',
    source: 'comment/index',
    styleDir: 'comment',
  },
  ConfigProvider: {
    local: 'default',
    exported: 'ConfigProvider',
    source: 'config-provider/index',
    styleDir: 'config-provider',
  },
  Empty: {
    local: 'default',
    exported: 'Empty',
    source: 'empty/index',
    styleDir: 'empty',
  },
  Result: {
    local: 'default',
    exported: 'Result',
    source: 'result/index',
    styleDir: 'result',
  },
  Descriptions: {
    local: 'default',
    exported: 'Descriptions',
    source: 'descriptions/index',
    styleDir: 'descriptions',
  },
  PageHeader: {
    local: 'default',
    exported: 'PageHeader',
    source: 'page-header/index',
    styleDir: 'page-header',
  },
  Space: {
    local: 'default',
    exported: 'Space',
    source: 'space/index',
    styleDir: 'space',
  },
};

export interface AntdResolverOptions {
  /**
   * @default 'A'
   */
  prefix?: string;

  /**
   * @default 'esm'
   */
  format?: 'esm' | 'cjs';

  /**
   * @default 'css'
   */
  importStyle?: boolean | 'css' | 'less';

  /**
   * @default false
   */
  specificImport?: boolean;

  /**
   * @default []
   */
  exclude?: string[];
}

/**
 * Resolver for ant-design-vue@1.x
 */
export function AntdResolver(options: AntdResolverOptions = {}): ComponentResolverObject {
  const { prefix = 'A', format = 'esm', importStyle = 'css', specificImport = false, exclude = [] } = options;

  return {
    type: 'component',
    resolve: (name: string) => {
      if (exclude.includes(name)) return;

      if (!name.startsWith(prefix)) return;
      const exportedName = name.slice(prefix.length);
      if (!exportedName) return;

      const exportedInfo = antdExportedMap[exportedName];
      if (!exportedInfo) return;

      const { local, exported, source, styleDir } = exportedInfo;
      const path = `ant-design-vue/${format === 'esm' ? 'es' : 'lib'}`;

      let importPath = path;
      let importName = exported;
      if (specificImport) {
        importPath = `${path}/${source}`;
        importName = local;
      }

      const stylePath = !importStyle
        ? undefined
        : !styleDir
          ? undefined
          : `${path}/${styleDir}/${importStyle === 'less' ? 'style/index' : 'style/css'}`;

      return {
        name: importName,
        from: importPath,
        sideEffects: stylePath,
      };
    },
  };
}
