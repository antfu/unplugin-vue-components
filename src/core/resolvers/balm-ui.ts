import { kebabCase } from '../utils'
import type { ComponentResolver, SideEffectsInfo } from '../../types'

const BalmUIDirectiveStyles = ['elevation', 'ripple', 'shape']

export interface BalmUIResolverOptions {
  /**
   * import style along with components
   *
   * @default false
   */
  importStyle?: false | 'css' | 'sass'
  /**
   * Use BalmUI source code.
   *
   * @default false
   */
  useSource?: boolean
}

function getFromSource(type: string, name: string, options: BalmUIResolverOptions): string {
  const unpluginList = type === 'component'
    ? ['UiSkeleton', 'UiEditor', 'UiTree', 'UiBottomNavigation', 'UiBottomSheet']
    : ['Anchor', 'Copy', 'Longpress']
  const isPlus = unpluginList.includes(name)

  return options.useSource
    ? (isPlus ? 'balm-ui/src/scripts/plus.js' : 'balm-ui/src/scripts/index.js')
    : (isPlus ? 'balm-ui/dist/balm-ui-plus.js' : 'balm-ui/dist/balm-ui.js')
}

function getSideEffects(type: string, name: string, options: BalmUIResolverOptions): SideEffectsInfo {
  const styleExtname = options.importStyle === 'sass' ? 'scss' : 'css'

  return `balm-ui/${type}s/${name}/${name}.${styleExtname}`
}

/**
 * Resolver for BalmUI
 *
 * @author @elf-mouse
 * @link https://material.balmjs.com/
 */
export function BalmUIResolver(options: BalmUIResolverOptions = {}): ComponentResolver[] {
  return [{
    type: 'component',
    resolve: (name: string) => {
      if (name.startsWith('Ui')) {
        const partialName = kebabCase(name.slice(2))

        return {
          name,
          from: getFromSource('component', name, options),
          sideEffects: options.importStyle
            ? getSideEffects('component', partialName, options)
            : undefined,
        }
      }
    },
  }, {
    type: 'directive',
    resolve: (name: string) => {
      const partialName = kebabCase(name)

      return {
        name: `v${name}`,
        from: getFromSource('directive', name, options),
        sideEffects: options.importStyle && BalmUIDirectiveStyles.includes(partialName)
          ? getSideEffects('directive', partialName, options)
          : undefined,
      }
    },
  }]
}
