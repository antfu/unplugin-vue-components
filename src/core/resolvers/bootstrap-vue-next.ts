import type { ComponentResolver } from '../../types'

export interface BootstrapVueNextResolverOptions {
  /**
   * Auto import for directives.
   *
   * @default true
   */
  directives?: boolean
}

/**
 * Resolver for BootstrapVue Next
 *
 * @link https://github.com/bootstrap-vue/bootstrap-vue-next
 */
export const BootstrapVueNextResolver = (_options: BootstrapVueNextResolverOptions = {}): Array<ComponentResolver> => {
  const options = { directives: true, ..._options }
  const resolvers: Array<ComponentResolver> = [{
    type: 'component',
    resolve: (name) => {
      if (name.match(/^B[A-Z]/))
        return { name, from: 'bootstrap-vue-next' }
    },
  }]

  if (options.directives) {
    resolvers.push({
      type: 'directive',
      resolve: (name) => {
        if (name.match(/^B[A-Z]/))
          return { name: `V${name}`, from: 'bootstrap-vue-next' }
      },
    })
  }

  return resolvers
}
