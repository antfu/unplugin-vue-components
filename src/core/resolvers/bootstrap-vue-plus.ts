import type { ComponentInfo, ComponentResolver, SideEffectsInfo } from '../../types'
import { getPkgVersion, kebabCase } from '../utils'

export interface BootstrapVuePlusResolverOptions {
  /**
   * import style css or sass with components
   *
   * @default 'css'
   */
  importStyle?: boolean | 'css' | 'sass'

  /**
   * use commonjs lib & source css or scss for ssr
   */
  ssr?: boolean

  /**
   * specify bootstrap-vue-plus version to load style
   *
   * @default installed version
   */
  version?: string

  /**
   * auto import for directives
   *
   * @default true
   */
  directives?: boolean

  /**
   * exclude component name, if match do not resolve the name
   */
  exclude?: RegExp
}

type BootstrapVuePlusResolverOptionsResolved = Required<Omit<BootstrapVuePlusResolverOptions, 'exclude'>> &
Pick<BootstrapVuePlusResolverOptions, 'exclude'>

function getSideEffects(dirName: string, options: BootstrapVuePlusResolverOptionsResolved): SideEffectsInfo | undefined {
  const { importStyle, ssr } = options
  const themeFolder = 'bootstrap-vue-plus/theme-chalk'
  const esComponentsFolder = 'bootstrap-vue-plus/es/components'

  if (importStyle === 'sass')
    return ssr ? `${themeFolder}/src/${dirName}.scss` : `${esComponentsFolder}/${dirName}/style/index`
  else if (importStyle === true || importStyle === 'css')
    return ssr ? `${themeFolder}/bv-${dirName}.css` : `${esComponentsFolder}/${dirName}/style/css`
}

function resolveComponent(name: string, options: BootstrapVuePlusResolverOptionsResolved): ComponentInfo | undefined {
  if (options.exclude && name.match(options.exclude))
    return

  if (!name.match(/^Bv[A-Z]/))
    return

  if (name.match(/^BvIcon.+/)) {
    return {
      name: name.replace(/^BvIcon/, ''),
      from: '@bootstrap-vue-plus/icons-vue',
    }
  }

  const partialName = kebabCase(name.slice(2))// ElTableColumn -> table-column
  const { ssr } = options

  return {
    name,
    from: `bootstrap-vue-plus/${ssr ? 'lib' : 'es'}`,
    sideEffects: getSideEffects(partialName, options),
  }
}

function resolveDirective(name: string, options: BootstrapVuePlusResolverOptionsResolved): ComponentInfo | undefined {
  if (!options.directives)
    return

  const directives: Record<string, { importName: string; styleName: string }> = {
    Message: { importName: 'BvMessage', styleName: 'message' },
  }

  const directive = directives[name]
  if (!directive)
    return

  const { ssr } = options

  return {
    name: directive.importName,
    from: `bootstrap-vue-plus/${ssr ? 'lib' : 'es'}`,
    sideEffects: getSideEffects(directive.styleName, options),
  }
}

/**
 * Resolver for Bootstrap Vue Plus
 *
 *
 * @author @see
 * @link https://bootstrap-vue-plus.com for bootstrap-vue-plus
 *
 */
export function BootstrapVuePlusResolver(
  options: BootstrapVuePlusResolverOptionsResolved = ({} as BootstrapVuePlusResolverOptionsResolved),
): ComponentResolver[] {
  let optionsResolved: BootstrapVuePlusResolverOptionsResolved

  async function resolveOptions() {
    if (optionsResolved)
      return optionsResolved
    optionsResolved = {
      ssr: false,
      version: await getPkgVersion('bootstrap-vue-plus', '0.0.1'),
      importStyle: 'css',
      directives: true,
      exclude: undefined,
      ...options,
    }
    return optionsResolved
  }

  return [
    {
      type: 'component',
      resolve: async (name: string) => {
        return resolveComponent(name, await resolveOptions())
      },
    },
    {
      type: 'directive',
      resolve: async (name: string) => {
        return resolveDirective(name, await resolveOptions())
      },
    },
  ]
}
