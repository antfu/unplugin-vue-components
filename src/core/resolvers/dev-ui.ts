import { kebabCase } from '../utils'
import type { ComponentResolver } from '../../types'

export interface DevResolverOptions {
  /**
   * bring in components and styles
   *
   * @default true
   */
  importStyle?: boolean

  /**
   * auto import for directives
   *
   * @default true
   */
  directives?: boolean
}

function getDefaultConf() {
  return {
    directive: true,
    importStyle: true,
  }
}

function getSideEffects(importName: string): string | undefined {
  return `vue-devui/${importName}/style.css`
}

function componentResolver(name: string, config: DevResolverOptions) { }

function directiveResolver(name: string, config: DevResolverOptions) { }

export function DevUiResolver(options: DevResolverOptions = {}): ComponentResolver {
  return {
    type: 'component',
    resolve: (name: string) => {
      if (name.match(/^D[A-Z]/)) {
        const config = { ...getDefaultConf(), ...options }

        name = name.slice(1)
        const _dir = kebabCase(name)

        return {
          importName: name,
          path: `vue-devui/${_dir}`,
          ...config.importStyle && { sideEffects: getSideEffects(_dir) },
        }
      }
    },
  }
}
