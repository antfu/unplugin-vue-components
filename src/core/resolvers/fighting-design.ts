import { isSSR, kebabCase } from '../utils'
import type { ComponentResolver } from '../../types'

const moduleType = isSSR ? 'lib' : 'es'

function getSideEffects(dirName: string): string {
  return `fighting-design/theme/${dirName}.css`
}

/**
 * Resolver for Fighting Design
 *
 * @author @Tyh2001
 * @link https://github.com/FightingDesign/fighting-design
 */
export function FightingDesignResolver(): ComponentResolver {
  return {
    type: 'component',
    resolve: (name: string) => {
      if (name.match(/^I[A-Z]/)) {
        return {
          name,
          from: `fighting-design/${moduleType}`,
          sideEffects: getSideEffects(kebabCase(name)),
        }
      }
    },
  }
}
