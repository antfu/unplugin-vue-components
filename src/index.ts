import type { Plugin } from 'vite'
import { createRollupPlugin } from './plugins/build'
import { createServerPlugin } from './plugins/server'
import { Options } from './types'
import { VueScriptTransformer } from './transforms/vueScript'
import { VueTemplateTransformer } from './transforms/vueTemplate'
import { SVGTransformer } from './transforms/svg'
import { Context } from './context'

const defaultOptions: Options = {
  dirs: 'src/components',
  extensions: 'vue',
  deep: true,

  directoryAsNamespace: false,
  globalNamespaces: [],

  alias: {},
  root: process.cwd(),
}

function VitePluginComponents(options: Partial<Options> = {}): Plugin {
  const resolvedOptions: Options = Object.assign({}, defaultOptions, options)
  const ctx: Context = new Context(resolvedOptions)

  return {
    configureServer: createServerPlugin(ctx),
    rollupInputOptions: {
      plugins: [
        createRollupPlugin(ctx),
      ],
    },
    transforms: [
      VueScriptTransformer(ctx),
      VueTemplateTransformer(ctx),
      SVGTransformer(ctx),
    ],
  }
}

export type { Options }
export default VitePluginComponents
