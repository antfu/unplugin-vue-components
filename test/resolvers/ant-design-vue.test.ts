import { describe, expect, it } from 'vitest'
import { AntDesignVueResolver } from '../../src/resolvers'
import { Context } from '../../src/core/context'

describe('Ant Design Vue Resolver', () => {
  it('components and directives should be transformed', async () => {
    const code = `
    import { defineComponent as _defineComponent } from "vue";\n' +
    'import { message } from "ant-design-vue";\n' +
    'import { ElMessage } from "element-plus";\n' +
    'const _sfc_main = /* @__PURE__ */ _defineComponent({\n' +
    '  __name: "App",\n' +
    '  setup(__props, { expose }) {\n' +
    '    expose();\n' +
    '    const test1 = () => {\n' +
    '      ElMessage.success("message");\n' +
    '    };\n' +
    '    const handleClick = () => {\n' +
    '      message.success("\\u6D4B\\u8BD5 jsx \\u4F7F\\u7528 a-button \\u548C mesage");\n' +
    '    };\n' +
    '    const __returned__ = { test1, handleClick };\n' +
    '    Object.defineProperty(__returned__, "__isScriptSetup", { enumerable: false, value: true });\n' +
    '    return __returned__;\n' +
    '  }\n' +
    '});\n' +
    'import { createTextVNode as _createTextVNode, resolveComponent as _resolveComponent, withCtx as _withCtx, createVNode as _createVNode, openBlock as _openBlock, createElementBlock as _createElementBlock, pushScopeId as _pushScopeId, popScopeId as _popScopeId } from "vue";\n' +
    'const _withScopeId = (n) => (_pushScopeId("data-v-7a7a37b1"), n = n(), _popScopeId(), n);\n' +
    'const _hoisted_1 = { class: "block" };\n' +
    'function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {\n' +
    '  const _component_a_button = _resolveComponent("a-button");\n' +
    '  return _openBlock(), _createElementBlock("div", _hoisted_1, [\n' +
    '    _createVNode(_component_a_button, { onClick: $setup.handleClick }, {\n' +
    '      default: _withCtx(() => [\n' +
    '        _createTextVNode(" \\u70B9\\u6211 ")\n' +
    '      ]),\n' +
    '      _: 1\n' +
    '      /* STABLE */\n' +
    '    })\n' +
    '  ]);\n' +
    '}\n' +
    'import "/Users/hejian/Documents/Github/unplugin-vue-components/examples/vite-vue3/src/App.vue?vue&type=style&index=0&scoped=7a7a37b1&lang.css";\n' +
    '_sfc_main.__hmrId = "7a7a37b1";\n' +
    'typeof __VUE_HMR_RUNTIME__ !== "undefined" && __VUE_HMR_RUNTIME__.createRecord(_sfc_main.__hmrId, _sfc_main);\n' +
    'import.meta.hot.accept((mod) => {\n' +
    '  if (!mod)\n' +
    '    return;\n' +
    '  const { default: updated, _rerender_only } = mod;\n' +
    '  if (_rerender_only) {\n' +
    '    __VUE_HMR_RUNTIME__.rerender(updated.__hmrId, updated.render);\n' +
    '  } else {\n' +
    '    __VUE_HMR_RUNTIME__.reload(updated.__hmrId, updated);\n' +
    '  }\n' +
    '});\n' +
    'import _export_sfc from "\\0plugin-vue:export-helper";\n' +
    'export default /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render], ["__scopeId", "data-v-7a7a37b1"], ["__file", "/Users/hejian/Documents/Github/unplugin-vue-components/examples/vite-vue3/src/App.vue"]]);\n
`

    const ctx = new Context({
      resolvers: [AntDesignVueResolver({})],
      transformer: 'vue3',
      directives: true,
    })
    ctx.sourcemap = false
    expect(await ctx.transform(code, '')).toMatchSnapshot()
  })
})
