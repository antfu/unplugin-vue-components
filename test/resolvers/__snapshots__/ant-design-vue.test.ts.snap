// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Ant Design Vue Resolver > components and directives should be transformed 1`] = `
{
  "code": "/* unplugin-vue-components disabled */import 'ant-design-vue/es/message/style/css';
import { Button as __unplugin_components_0 } from 'ant-design-vue/es';import 'ant-design-vue/es/button/style/css';
    import { defineComponent as _defineComponent } from \\"vue\\";
' +
    'import { message } from \\"ant-design-vue\\";
' +
    'import { ElMessage } from \\"element-plus\\";
' +
    'const _sfc_main = /* @__PURE__ */ _defineComponent({
' +
    '  __name: \\"App\\",
' +
    '  setup(__props, { expose }) {
' +
    '    expose();
' +
    '    const test1 = () => {
' +
    '      ElMessage.success(\\"message\\");
' +
    '    };
' +
    '    const handleClick = () => {
' +
    '      message.success(\\"\\\\u6D4B\\\\u8BD5 jsx \\\\u4F7F\\\\u7528 a-button \\\\u548C mesage\\");
' +
    '    };
' +
    '    const __returned__ = { test1, handleClick };
' +
    '    Object.defineProperty(__returned__, \\"__isScriptSetup\\", { enumerable: false, value: true });
' +
    '    return __returned__;
' +
    '  }
' +
    '});
' +
    'import { createTextVNode as _createTextVNode, resolveComponent as _resolveComponent, withCtx as _withCtx, createVNode as _createVNode, openBlock as _openBlock, createElementBlock as _createElementBlock, pushScopeId as _pushScopeId, popScopeId as _popScopeId } from \\"vue\\";
' +
    'const _withScopeId = (n) => (_pushScopeId(\\"data-v-7a7a37b1\\"), n = n(), _popScopeId(), n);
' +
    'const _hoisted_1 = { class: \\"block\\" };
' +
    'function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
' +
    '  const _component_a_button = __unplugin_components_0;
' +
    '  return _openBlock(), _createElementBlock(\\"div\\", _hoisted_1, [
' +
    '    _createVNode(_component_a_button, { onClick: $setup.handleClick }, {
' +
    '      default: _withCtx(() => [
' +
    '        _createTextVNode(\\" \\\\u70B9\\\\u6211 \\")
' +
    '      ]),
' +
    '      _: 1
' +
    '      /* STABLE */
' +
    '    })
' +
    '  ]);
' +
    '}
' +
    'import \\"/Users/hejian/Documents/Github/unplugin-vue-components/examples/vite-vue3/src/App.vue?vue&type=style&index=0&scoped=7a7a37b1&lang.css\\";
' +
    '_sfc_main.__hmrId = \\"7a7a37b1\\";
' +
    'typeof __VUE_HMR_RUNTIME__ !== \\"undefined\\" && __VUE_HMR_RUNTIME__.createRecord(_sfc_main.__hmrId, _sfc_main);
' +
    'import.meta.hot.accept((mod) => {
' +
    '  if (!mod)
' +
    '    return;
' +
    '  const { default: updated, _rerender_only } = mod;
' +
    '  if (_rerender_only) {
' +
    '    __VUE_HMR_RUNTIME__.rerender(updated.__hmrId, updated.render);
' +
    '  } else {
' +
    '    __VUE_HMR_RUNTIME__.reload(updated.__hmrId, updated);
' +
    '  }
' +
    '});
' +
    'import _export_sfc from \\"\\\\0plugin-vue:export-helper\\";
' +
    'export default /* @__PURE__ */ _export_sfc(_sfc_main, [[\\"render\\", _sfc_render], [\\"__scopeId\\", \\"data-v-7a7a37b1\\"], [\\"__file\\", \\"/Users/hejian/Documents/Github/unplugin-vue-components/examples/vite-vue3/src/App.vue\\"]]);

",
}
`;
